### PR TITLE
Implement NULL metadata relaying in the SDK/CLI

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Add NULL self relay to core app
- Add self relay to send message command

### Drive-by changes

- Migrate to assert util in ISM Factory
- Default to `HYP_KEY` in shared key option

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3608
- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3609
- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3452

### Backward compatibility

Yes

### Testing

Manual

https://github.com/hyperlane-xyz/hyperlane-monorepo/assets/3020995/e44683c3-bca4-41d5-a4c7-45718f05edf4

https://github.com/hyperlane-xyz/hyperlane-monorepo/assets/3020995/50565be1-d266-4d2d-ae27-5fbb5539b93f

### CODE REVIEW

File: typescript/cli/src/send/message.ts [hunk 1]
```diff
@@ -146,6 +146,9 @@ async function executeDelivery({
     logBlue(`Sent message from ${origin} to ${recipient} on ${destination}.`);
     logBlue(`Message ID: ${message.id}`);
     log(`Message: ${JSON.stringify(message)}`);
+
+    await core.relayMessage(message);
```
Comment (jmrossy) [inline typescript/cli/src/send/message.ts:150]: Should we check if the destination is a core chain and let regular relay handle those?

File: typescript/sdk/src/core/HyperlaneCore.ts [hunk 2]
```diff
@@ -85,6 +89,50 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
...
+    message: DispatchedMessage,
+  ): Promise<ethers.ContractReceipt> {
+    const destinationChain = this.getDestination(message);
+    const mailbox = this.contractsMap[destinationChain].mailbox;
+    const ismConfig = await this.getRecipientIsmConfig(message);
+    if (typeof ismConfig === 'string') {
+      throw new Error(
+        `No ISM config found for recipient ${message.parsed.recipient}`,
+      );
+    }
```
Comment (jmrossy) [inline typescript/sdk/src/core/HyperlaneCore.ts:117]: Does a string config mean it's an address? The error message is confusing me a bit here

File: typescript/sdk/src/core/HyperlaneCore.ts [hunk 3]
```diff
@@ -85,6 +89,50 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
...
+    // TODO: implement metadata builders
+    let metadata: string;
+    const moduleType = ismTypeToModuleType(ismConfig.type);
+    switch (moduleType) {
+      case ModuleType.NULL:
+        metadata = '0x';
+        break;
+      default:
+        throw new Error(`Unsupported module type ${moduleType}`);
+    }
```
Comment (jmrossy) [inline typescript/sdk/src/core/HyperlaneCore.ts:128]: In my ongoing war against switch statements, the worst control flow tool ever invented, I'll complain one more time:
is this really better than
```ts
if ( moduleType === ModuelType.NULL) metadata = '0x';
else throw new Error(`Unsupported module type ${moduleType}`);
```

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 4]
```diff
@@ -176,6 +177,21 @@ export class HyperlaneIsmFactory extends HyperlaneApp<ProxyFactoryFactories> {
...
         });
         break;
+      case IsmType.TRUSTED_RELAYER:
+        if (!this.deployer) {
+          throw new Error(`HyperlaneDeployer must be set to deploy ${ismType}`);
+        } else if (!mailbox) {
+          throw new Error(
+            `Mailbox address is required for deploying ${ismType}`,
+          );
+        }
```
Comment (jmrossy) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts:187]: NIt: Could be `assert(this.deployer, `HyperlaneDeployer must be set to deploy ${ismType})`.
See assert in our utils package

**Review by jmrossy — APPROVED** [review]
Nice work here! Impressed with how fast you put this together. 
I will say though, I think our M.O. should be for all new SDK features to be multi-protocol first but I'll leave that decision to you

File: typescript/sdk/src/core/HyperlaneCore.ts (hunk 3 above)
Comment (nbayindirli) [reply typescript/sdk/src/core/HyperlaneCore.ts:128]: +1, so long as `ModuleType.NULL` comes first in the if statement (purely cosmetic)

LOGAF super low on switch statements

File: typescript/cli/src/config/ism.ts [hunk 5]
```diff
@@ -159,36 +165,34 @@ export async function createIsmConfig(
...
       },
       {
         value: IsmType.FALLBACK_ROUTING,
-        name: IsmType.FALLBACK_ROUTING,
         description:
           "You can specify ISM type for specific chains you like and fallback to mailbox's default ISM for other chains via DefaultFallbackRoutingISM",
       },
       {
         value: IsmType.AGGREGATION,
-        name: IsmType.AGGREGATION,
```
Comment (paulbalaji) [inline typescript/cli/src/config/ism.ts:185]: thanks for the drive-by

File: typescript/cli/src/send/message.ts (hunk 1 above)
Comment (paulbalaji) [reply typescript/cli/src/send/message.ts:150]: imo self-relay is analogous to `relay --force` so it makes sense to bypass regular relaying

File: typescript/sdk/src/core/HyperlaneCore.ts [hunk 6]
```diff
@@ -85,6 +89,50 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
...
+    return destinationMailbox.mailbox.recipientIsm(ethAddress);
+  }
+
+  async getRecipientIsmConfig(
+    message: DispatchedMessage,
+  ): Promise<WithAddress<IsmConfig>> {
+    const destinationChain = this.getDestination(message);
+    const ismReader = new EvmIsmReader(this.multiProvider, destinationChain);
+    const address = await this.getRecipientIsmAddress(message);
+    return ismReader.deriveIsmConfig(address);
```
Comment (paulbalaji) [inline typescript/sdk/src/core/HyperlaneCore.ts:104]: image

File: typescript/sdk/src/core/HyperlaneCore.ts (hunk 2 above)
Comment (paulbalaji) [reply typescript/sdk/src/core/HyperlaneCore.ts:117]: does `getRecipientIsmConfig` actually return a type `string`? I would've thought that with the type `WithAddress<IsmConfig>`, we'd get back an object instead of just the string i.e.
```
{
    string: '0xsdsd',
    address: '0xsdsd',
}
```

curious if you hit a problem when running that led to this check

File: typescript/sdk/src/core/HyperlaneCore.ts (hunk 3 above)
Comment (paulbalaji) [reply typescript/sdk/src/core/HyperlaneCore.ts:128]: rossy v switch statements
image

File: solidity/contracts/isms/TrustedRelayerIsm.sol [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import {IInterchainSecurityModule} from "../interfaces/IInterchainSecurityModule.sol";
+import {Message} from "../libs/Message.sol";
+import {Mailbox} from "../Mailbox.sol";
+
+contract TrustedRelayerIsm is IInterchainSecurityModule {
+    using Message for bytes;
+
+    uint8 public moduleType = uint8(Types.NULL);
+    Mailbox public immutable mailbox;
+    address public immutable trustedRelayer;
```
Comment (paulbalaji) [inline solidity/contracts/isms/TrustedRelayerIsm.sol:14]: update `deriveNullConfig` in the ISM reader pls, if contract has `trustedRelayer` property then you can infer it's a `TrustedRelayerIsm`

File: typescript/sdk/src/core/HyperlaneCore.ts (hunk 2 above)
Comment (yorhodes) [reply typescript/sdk/src/core/HyperlaneCore.ts]: changed your return type to `Exclude`

File: typescript/sdk/src/core/HyperlaneCore.ts (hunk 3 above)
Comment (yorhodes) [reply typescript/sdk/src/core/HyperlaneCore.ts]: I got rid of the mutation but keeping the switch case after chatting offline :P

File: typescript/cli/src/send/message.ts (hunk 1 above)
Comment (yorhodes) [reply typescript/cli/src/send/message.ts]: added an explicit --selfrelay flag

File: typescript/sdk/src/ism/read.ts [hunk 8]
```diff
@@ -184,9 +192,27 @@ export class EvmIsmReader implements IsmReader {
 
   async deriveNullConfig(
     address: Address,
-  ): Promise<
-    WithAddress<PausableIsmConfig | TestIsmConfig | OpStackIsmConfig>
-  > {
+  ): Promise<WithAddress<NullIsmConfig>> {
+    // if it has trustedRelayer() property --> TrustedRelayer ISM
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/read.ts:196]: thanks for adding! might need to also mock `TrustedRelayerIsm__factory` in the associated test to make CI happy

**Review by paulbalaji — APPROVED** [review]
super stoked with how quick this came together

File: typescript/cli/src/commands/send.ts [hunk 9]
```diff
@@ -85,6 +93,7 @@ const messageCommand: CommandModule = {
       messageBody: ethers.utils.hexlify(ethers.utils.toUtf8Bytes(messageBody)),
       timeoutSec,
       skipWaitForDelivery,
+      selfRelay,
```
Comment (nbayindirli) [inline typescript/cli/src/commands/send.ts:96]: Strongly in favor of this being '`--self-relay`' via CLI

**Review by nbayindirli — CHANGES_REQUESTED** [review]

File: typescript/cli/src/commands/send.ts [hunk 10]
```diff
@@ -32,6 +32,7 @@ export const selfrelay: Options = {
   type: 'boolean',
   description: 'Relay message on destination chain',
   default: false,
+  alias: ['s', 'sr'],
```
Comment (nbayindirli) [inline typescript/cli/src/commands/send.ts:35]: Thank you!

**Review by nbayindirli — APPROVED** [review]
🚢
